pkg/s3select/input.go

```diff
@@ -0,0 +1,322 @@
+		options: opts,
+		reader:  csv.NewReader(myReader),
+		stats:   progress,
+	}
+	reader.firstRow = nil
+
+	reader.reader.FieldsPerRecord = -1
+	if reader.options.FieldDelimiter != "" {
+		reader.reader.Comma = rune(reader.options.FieldDelimiter[0])
+	}
+
+	if reader.options.Comments != "" {
+		reader.reader.Comment = rune(reader.options.Comments[0])
+	}
+
+	// QuoteCharacter - " (defaulted currently)
+	reader.reader.LazyQuotes = true
+
+	if err := reader.readHeader(); err != nil {
+		return nil, err
+	}
+
+	return reader, nil
+}
+
+// ReadRecord reads a single record from the . Always returns successfully.
+// If the record is empty, an empty []string is returned.
+// Record expand to match the current row size, adding blank fields as needed.
+// Records never return less then the number of fields in the first row.
+// Returns nil on EOF
+// In the event of a parse error due to an invalid record, it is logged, and
+// an empty []string is returned with the number of fields in the first row,
+// as if the record were empty.
+//
+// In general, this is a very tolerant of problems  reader.
+func (reader *Input) ReadRecord() []string {
+	var row []string
+	var fileErr error
+
+	if reader.firstRow != nil {
+		row = reader.firstRow
+		reader.firstRow = nil
+		return row
+	}
+
+	row, fileErr = reader.reader.Read()
+	emptysToAppend := reader.minOutputLength - len(row)
+	if fileErr == io.EOF {
+		return nil
+	} else if _, ok := fileErr.(*csv.ParseError); ok {
+		emptysToAppend = reader.minOutputLength
+	}
+
+	if emptysToAppend > 0 {
+		for counter := 0; counter < emptysToAppend; counter++ {
+			row = append(row, "")
+		}
+	}
+
+	return row
+}
+
+// convertMySQL Replaces double quote escape for column names with backtick for
+// the MySQL parser
+func convertMySQL(random string) string {
+	return strings.Replace(random, "\"", "`", len(random))
+}
+
+// readHeader reads the header into the header variable if the header is present
+// as the first row of the csv
+func (reader *Input) readHeader() error {
+	var readErr error
+	if reader.options.HasHeader {
+		reader.firstRow, readErr = reader.reader.Read()
+		if readErr != nil {
+			return ErrCSVParsingError
+		}
+		reader.header = reader.firstRow
+		reader.firstRow = nil
+		reader.minOutputLength = len(reader.header)
+	}
+	return nil
+}
+
+// createStatXML is the function which does the marshaling from the stat
+// structs into XML so that the progress and stat message can be sent
+func (reader *Input) createStatXML() string {
+	if reader.options.CompressedGzip {
+		reader.stats.BytesProcessed += processSize(reader.header)
+	}
+	if !reader.options.CompressedGzip {
+		reader.stats.BytesProcessed = reader.options.StreamSize
+		reader.stats.BytesScanned = reader.stats.BytesProcessed
+	}
+	statXML := &stats{
+		BytesScanned:   reader.stats.BytesScanned,
+		BytesProcessed: reader.stats.BytesProcessed,
+		BytesReturned:  reader.stats.BytesReturned,
+	}
+	out, _ := xml.MarshalIndent(statXML, "", "   ")
```

Comment:
Don't ignore the error here we can send an error back to the client here, also do we need to Indent?